Add API for Stackmembers

stveit opened the pull request:
Adds PUT and GET API for Stackmembers. Uses pydantic for input validation. 

It does not have a Flask model, as I made a pydantic model instead. But a Flask model may be required for the automated API documentation, I think?

Device API tests are rewritten to not rely on data pre-existing in the database. 

Function added to session for wiping the database between tests. This is needed, as the API uses different sessions from the tests, so if i want to see if stuff was actually added to the database I have to fully commit stuff to the db and then check. This kinda overlaps with the test_sqla_session, so device db tests can be rewritten to use clear_db instead of test_sqla_session to remove this redundancy.

Currently hardware_id accepts numbers as well as str. This was not initally intentional, as it turned out that using `str` in pydantic also allows numbers (ints and floats i tested with), ad automatically converts them to strings. Probably holds true for other data types as well. This can be fixed by using StrictStr or something akin to that. I havent changed it yet as maybe there is a reason to allow stuff that can be converted to a string.

Probably more stuff to discuss that i havent considered

indy-independence commented:
Flask model is required to get the automated swagger thing working yes
Then we also need manual documentation in docs/ rst format

indy-independence commented on docs/apiref/devices.rst:
```diff
@@ -364,3 +364,59 @@ NMS CA (specified in api.yml).
+
+
+Show stackmembers
+--------------
```
too few ---

indy-independence commented on docs/apiref/devices.rst:
```diff
@@ -364,3 +364,59 @@ NMS CA (specified in api.yml).
+  }  
+
+Set stackmembers
+--------------
```
too few --

indy-independence commented on src/cnaas_nms/api/device.py:
```diff
@@ -1,6 +1,7 @@
 import json
 import datetime
 from typing import Optional
+from typing_extensions import Required
```
unused import?

stveit replied:
yeah, auto imported by mistake ..

indy-independence commented on src/cnaas_nms/api/device.py:
```diff
@@ -127,6 +128,15 @@
+
+stackmembers_model = device_api.model('stackmembers', {
+    'stackmembers': fields.List(fields.Nested(stackmember_model)),
+})
```
My editor complains about pep8 "two whitespace before defining class" after this

stveit replied:
Turned on auto removing of whitespaces now (should have done this months ago ..) and found whitespaces elsewhere but not there, so hopefully its somehow fixed

indy-independence replied:
Ah sorry I was unclear, my editor wants to have two newlines before defining a new class because of pep8, so it's too few newlines/whitespace not too many

indy-independence reviewed (changes requested):
Looking good! I think we should change API so that it refers to devices by hostname instead of device_id if possible. Most places use hostname except for init and put where you actually specify a new/modified hostname so it makes sense to use id there instead

stveit commented:
did some unecesary force push, but i think i fixed it by force pushing again and resetting it. If not, just force pull, im guessing im the only one making commits here.

Fixed the things you mentioned

indy-independence reviewed (commented):
Ok great! I just realized docs has to be updated a bit also, maybe you're already working on it

stveit commented:
Might be the env variables that are set at the start of integrationtests.sh the USERNAME_* stuff. The new tests set these as well, but to different values

stveit commented:
or maybe not, thought the postgres username/password was set there but its not

stveit commented:
Could maybe be Alembic issues related to the new stackmember model? If database is not updated with newest revision it will fail

indy-independence commented:
It says in docker logs that it does:
2021-12-01 15:25:34,272 DEBG 'uwsgi' stdout output:
INFO  [alembic.runtime.migration] Running upgrade 9d01bce3c835 -> b7629362583c, Add table for stack members

But it's also complaining about:
ImportError: cannot import name 'Required' from 'typing_extensions' (/opt/cnaas/venv/lib/python3.7/site-packages/typing_extensions.py)
which is not in the current code so seems like it's using a bit old code for some reason..

stveit commented:
I did do some force pushing prior to fixing the import, could this have broken something that is not able to pull and update now? Or maybe a rebuild of the image is required? just spitballing

indy-independence commented:
I tried running the integrationtests.sh on my laptop now, did you ever get that running at uninett btw? With virtual vEOS?

Some unit test breaks from the changes here, but in total the coverage percentage is still the same or better than it was before :joy: 

Some outputs:
```
======================================================================
ERROR: test_get_stackmembers (cnaas_nms.db.tests.test_device.DeviceTests)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/opt/cnaas/venv/lib/python3.7/site-packages/sqlalchemy/engine/base.py", line 1277, in _execute_context
    cursor, statement, parameters, context
  File "/opt/cnaas/venv/lib/python3.7/site-packages/sqlalchemy/engine/default.py", line 593, in do_execute
    cursor.execute(statement, parameters)
psycopg2.errors.UniqueViolation: duplicate key value violates unique constraint "stackmember_device_id_hardware_id_key"
DETAIL:  Key (device_id, hardware_id)=(12, FO64534) already exists.
```

Then I think the management domain tests fails because there are no dist devices:
test_find_free_mgmt_ip
test_find_mgmt_omain
test_find_mgmtdomain_by_ip
test_initcheck_distdevice

test_get_inventory might have been broken since before, but complains on:
len(result['hosts'].items()))
AssertionError: 1 not less than or equal to 0

If you can get integrationtests.sh running at uninett and somehow fix those 4-5 unittests I think we should be good

indy-independence commented:
Maybe teardown/clean_db() function can reset database to "known state" with eosdist1/2 and eosaccess so other tests can run instead of leaving it empty?

stveit commented:
I have not gotten integrationtests.sh working yet sadly, but i am trying now. 
I've tried now on a completely fresh set up to run the integratintests.sh on develop, but i get bad status code 500. Are the VMs required for the tests to work at all? I was hoping if the VMs are not setup then just the tests requiring the VMs would fail, but I would still be able to run most tests

indy-independence commented:
I don't think it will do any of the integrationtests if you don't have at least eosdist1 and eosaccess VMs running

stveit commented:
When you run the integration test for this API, do you just change the BUILDBRANCH/GITREPO_BASE in docker-compose.yml or is it some magic in integrationtests.sh?

indy-independence commented:
First I run docker build like this:
docker-compose build --pull --build-arg BUILDBRANCH=origin/pr/234 cnaas_api
then do:
cd test/
./integrationtests.sh

stveit commented:
Im having issues with the coverage stuff, seems to fail during exit of the tests or something. Have you had this before?

```
Ran 54 tests in 0.743s

OK
Traceback (most recent call last):
  File "/opt/cnaas/venv/bin/nosetests", line 8, in <module>
    sys.exit(run_exit())
  File "/opt/cnaas/venv/lib/python3.7/site-packages/nose/core.py", line 121, in __init__
    **extra_args)
  File "/usr/lib/python3.7/unittest/main.py", line 101, in __init__
    self.runTests()
  File "/opt/cnaas/venv/lib/python3.7/site-packages/nose/core.py", line 207, in runTests
    result = self.testRunner.run(self.test)
  File "/opt/cnaas/venv/lib/python3.7/site-packages/nose/core.py", line 68, in run
    self.config.plugins.finalize(result)
  File "/opt/cnaas/venv/lib/python3.7/site-packages/nose/plugins/manager.py", line 99, in __call__
    return self.call(*arg, **kw)
  File "/opt/cnaas/venv/lib/python3.7/site-packages/nose/plugins/manager.py", line 167, in simple
    result = meth(*arg, **kw)
  File "/opt/cnaas/venv/lib/python3.7/site-packages/nose/plugins/testid.py", line 166, in finalize
    fh = open(self.idfile, 'wb')
PermissionError: [Errno 13] Permission denied: '/opt/cnaas/venv/cnaas-nms/src/.noseids'
Traceback (most recent call last):
  File "/opt/cnaas/venv/bin/nosetests", line 8, in <module>
    sys.exit(run_exit())
  File "/opt/cnaas/venv/lib/python3.7/site-packages/nose/core.py", line 121, in __init__
    **extra_args)
  File "/usr/lib/python3.7/unittest/main.py", line 100, in __init__
    self.parseArgs(argv)
  File "/opt/cnaas/venv/lib/python3.7/site-packages/nose/core.py", line 145, in parseArgs
    self.config.configure(argv, doc=self.usage())
  File "/opt/cnaas/venv/lib/python3.7/site-packages/nose/config.py", line 346, in configure
    self.plugins.configure(options, self)
  File "/opt/cnaas/venv/lib/python3.7/site-packages/nose/plugins/manager.py", line 284, in configure
    cfg(options, config)
  File "/opt/cnaas/venv/lib/python3.7/site-packages/nose/plugins/manager.py", line 99, in __call__
    return self.call(*arg, **kw)
  File "/opt/cnaas/venv/lib/python3.7/site-packages/nose/plugins/manager.py", line 167, in simple
    result = meth(*arg, **kw)
  File "/opt/cnaas/venv/lib/python3.7/site-packages/nose/plugins/cover.py", line 155, in configure
    self.coverInstance.load()
  File "/opt/cnaas/venv/lib/python3.7/site-packages/coverage/control.py", line 425, in load
    self._data.read()
  File "/opt/cnaas/venv/lib/python3.7/site-packages/coverage/sqldata.py", line 742, in read
    with self._connect():       # TODO: doesn't look right
  File "/opt/cnaas/venv/lib/python3.7/site-packages/coverage/sqldata.py", line 299, in _connect
    self._create_db()
  File "/opt/cnaas/venv/lib/python3.7/site-packages/coverage/sqldata.py", line 248, in _create_db
    with db:
  File "/opt/cnaas/venv/lib/python3.7/site-packages/coverage/sqldata.py", line 1026, in __enter__
    self._connect()
  File "/opt/cnaas/venv/lib/python3.7/site-packages/coverage/sqldata.py", line 1008, in _connect
    self.con = sqlite3.connect(filename, check_same_thread=False)
sqlite3.OperationalError: unable to open database file
cp: cannot stat '.coverage': No such file or directory
```

stveit commented:
Wasnt able to get the vm stuff working yet, but ive pushed a messy commit that should make it so the device tests run without messing up anything in the database.

wouter1975 commented on src/cnaas_nms/api/device.py:
```diff
@@ -123,6 +127,16 @@
+stackmember_model = device_api.model('stackmember', {
+    'member_no': fields.Integer(required=False),
+    'hardware_id': fields.String(required=True),
+    'prioity_id': fields.Integer(required=False)
```
typo: priority_id
